osu.Game/Screens/Play/SubmittingPlayer.cs

```diff
@@ -176,13 +178,19 @@ public override bool OnExiting(ScreenExitEvent e)
         [CanBeNull]
         protected abstract APIRequest<APIScoreToken> CreateTokenRequest();
 
+        /// <summary>
+        /// Constructs a <see cref="ScoreToken"/> for later use
+        /// from the <see cref="APIScoreToken"/> returned by web.
+        /// </summary>
+        protected abstract ScoreToken RetrieveScoreToken(APIScoreToken token);
```

Comment:
I generally wanted to avoid creating this method and try to constrain this in `CreateTokenRequest()`, but the C# type system makes this annoying.

tl;dr: I would ideally do

```csharp
protected abstract APIRequest<IAPIScoreToken> CreateTokenRequest()
```

with two implementations of `IAPIScoreToken` for solo and for multi. However the C# type system makes this extra annoying, since you can only have variance on interface generics, so that'd mean I'd have to do `IAPIRequest<T>` first, then probably something like `IAPIScoreTokenRequest<T>`, only then can it have `IAPIScoreToken` in the generic...... I don't think so. Unless you consider that better I guess.